File: C++/Clone-Graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    UndirectedGraphNode *helper(UndirectedGraphNode* node, 
        unordered_map<UndirectedGraphNode*, UndirectedGraphNode*>& hash) {
        if (hash.find(node) != hash.end()) return hash[node];
        UndirectedGraphNode *cur = new UndirectedGraphNode(node->label);
        hash.insert(make_pair(node, cur));         
        for (int i = 0; i < (node->neighbors).size(); i ++) {
            (cur->neighbors).push_back(helper((node->neighbors)[i], hash));
        }
        return cur;
    }
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if (node == NULL) return node;
        unordered_map<UndirectedGraphNode*, UndirectedGraphNode*> hash;
        helper(node, hash);
        return hash[node];
    }
};

File: Python/Clone-Graph.py
# Definition for a undirected graph node
class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if node == None:
            return None
        d = {}
        seq = []
        que = []
        res = UndirectedGraphNode(node.label)
        seq.append(node)
        que.append(res)
        d[node.label] = res
        while len(seq) > 0:
            cnt = seq.pop(0)
            ans = que.pop(0)
            for item in cnt.neighbors:
                if item.label not in d:
                    cur = UndirectedGraphNode(item.label)
                    d[item.label] = cur
                    ans.neighbors.append(cur)
                    seq.append(item)
                    que.append(cur)
                else:
                    ans.neighbors.append(d[item.label])
        return res
